fetchai/agents-tac
# Rejoin procedure

This PR implements the rejoin procedure for a TAC Agent. 

This feature is useful in the case when an agent, after correctly registered, stops. It might happen during the registration phase or during the competition phase. In both cases, the agent does the following:

- Search for controllers
- Once a controller has been found, the agent sends a `GetStateUpdate` message
- Once received a `StateUpdate` message, the `on_state_update` method is called, which resets the `GameInstance` object, updating the current state of the agent in the TAC instance.

## How to test

- (only once) reinstall the `tac` package
```
python3 setup.py install --force
```
- generate a private key:
```
python3 scripts/generate_private_key.py private_key.pem
```

- After `docker-compose build` (only once), run a sandbox with only two agents:
```
NB_AGENTS=2 NB_BASELINE_AGENTS=1 docker-compose up
```

- During registration time, run the following command:
```
python3 templates/v2/basic.py --name my-agent --private-key private_key.pem
```

- *Before* the TAC starts (just for the sake of the test, this constraint won't be needed in the future), `CTRL-C` the script
- Wait until the TAC starts (you can inspect it from the logs and/or from http://localhost:8097)
- run again the same command, but with the `--rejoin` flag:
```
python3 templates/v2/basic.py --name my-agent --private-key private_key.pem --rejoin
```
- Now you should see your agent that, despite restarted, is able to trade with the other agents.
